src/Sylius/Bundle/ApiBundle/Validator/GatewayConfig/GatewayConfigValidator.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\GatewayConfig;
+
+use Sylius\Bundle\ApiBundle\Validator\Constraints\GatewayConfig;
+use Sylius\Bundle\PayumBundle\Model\GatewayConfigInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+final class GatewayConfigValidator extends ConstraintValidator
+{
+    public const OFFLINE_GATEWAY_FACTORY = 'offline';
+
+    /** @param array<string, string> $factoryNames */
+    public function __construct(private array $factoryNames)
+    {
+    }
+
+    /** @param GatewayConfigInterface $value */
+    public function validate($value, Constraint $constraint): void
+    {
+        if (!$constraint instanceof GatewayConfig) {
+            throw new UnexpectedTypeException($constraint, GatewayConfig::class);
+        }
+
+        if (!$value instanceof GatewayConfigInterface) {
+            throw new UnexpectedValueException($value, GatewayConfigInterface::class);
+        }
+
+        /** @var string|null $factoryName */
+        $factoryName = $value->getFactoryName();
+
+        if ($factoryName === self::OFFLINE_GATEWAY_FACTORY || $factoryName === null) {
+            return;
+        }
+
+        if (!in_array($factoryName, array_keys($this->factoryNames), true)) {
+            $this->context->buildViolation($constraint->invalidGatewayFactory)
+                ->atPath('factoryName')
+                ->setParameter('{{ available_factories }}', implode(', ', array_keys($this->factoryNames)))
+                ->addViolation()
+            ;
+
+            return;
+        }
+
+        /** @var string[] $groups */
+        $groups = array_merge($constraint->groups, [$factoryName]);
+        $validator = $this->context->getValidator()->inContext($this->context);
```

Comment:
How about making a custom validator class per factory type?

> PaypalGatewayConfigValidator
> StripeGatewayConfigValidator
> etc

It will be a better extension point for framework end-users and example for plugin authors.

Current approach is original, but feels more app-level.

WDYT?

Reply:
/cc @Wojdylak @jakubtobiasz

Reply:
I was thinking more about separating XML files with constraints per factory type (e.g., `GatewayConfigStripeCheckout.xml`,` GatewayConfigPaypalExpressCheckout.xml`). IMO, creating custom validators that will do the same thing as Symfony constraints is unnecessary.

Reply:
@Wojdylak 
With XML approach,

Pros
- No custom validator per factory, which doesn't require custom group and can check internally if it can be applied to current factory. :slightly_smiling_face: 

Cons
- It's a sort of convention to name file with class name, that it will validate.
- Some `groups` magic still needed, which preferably should happen in each plugin, not at sylius/sylius. (Like a compiler pass or other "official" way to add/modify configuration)
- What about this? https://github.com/Sylius/Sylius/pull/15549#discussion_r1407035681 I didn't check, but have a feeling it works now only due some internal validator's recursive call check, which is not a good thing to base new code on.

Reply:
If I understand correctly, you are suggesting something like what is [here](https://github.com/Sylius/Sylius/blob/1.13/src/Sylius/Bundle/PromotionBundle/Validator/CatalogPromotionActionValidator.php#L53C1-L53C1). IMO, it seems like a similar mechanism because it's `sylius/sylius` that is responsible for triggering the validation on the promotion action.

https://github.com/Sylius/Sylius/pull/15549#discussion_r1407035681
I also don't know exactly how it works, and I need to look into it. But for me, the advantage is that the `symfony/validator`  do it for me.

Reply:
Here is another alternative

https://github.com/Sylius/Sylius/blob/75fd302e6038ac08ce74645b46970c751d070d70/src/Sylius/Component/Attribute/AttributeType/TextAttributeType.php#L36-L54

Reply:
Okey, so you don't want this mechanism to always trigger validation with that group without the person who added a new factory in the plugin knowing?

This example is simple because we're only checking `NotBlank`. For me, a big downside is that if I have a more complex case where I want to check `NotBlank`, `Type(int)`, `Range` of first field and `NotBlank`, Type(string)` of second field then I will have to create custom logic.

Reply:
> Okey, so you don't want this mechanism to always trigger validation with that group without the person who added a new factory in the plugin knowing?

Yes, exactly.

> This example is simple because we're only checking NotBlank. For me, a big downside is that if I have a more complex case where I want to check

It's true, but if catalog promotion already solve it in some way, then IMO we should re-use approach.

Reply:
@diimpp 
This is just a suggestion. What if we added a new parameter `sylius.gateway_factory.validators` where the key will be the name of the factory and the value will be the group that the validator should trigger. We would inject this parameter into the `GatewayConfigValidator` and if it doesn't find the key, it won't trigger any validation. Someone in the plugin will have to add an new value to this parameter to trigger validation."